index.rst

```diff
@@ -1,59 +1,367 @@
+
+The default (POSIX) datastore will write datasets with templates that begin with the ``RUN`` name, resulting in e.g. the datasets of per-instrument ``RUN`` collections landing in ``/datasets/repo_<YYYYMMDD>/<instrument>/`` and per-user ``RUN`` collections landing in ``/datasets/repo_<YYYYMMDD>/u/<user>``.
+Users are discouraged from inspecting these directories (as this will be at least quite different in the IDF or other future cloud-based datastores), and *strongly* discouraged from modifying them in any way other than via middleware tools.
+In many cases, write access will actually be prohibited (see :ref:`access-controls`).
+
+When migrations are necessary due to changes in the repository format (something that is *always* preceded by an RFC with explicit CCB approval), a new ``repo_<YYYYMMDD>`` directory and database/schema pair will be created, and files will shared via hard links until/unless the old repository is retired.
+
+.. note::
+
+   **TODO**: Are hard links viable here from a sysadmin/GPFS perspective?
+   They certainly would make things easier, because they'd let each repository have its own file without actual duplicationof storage.  Note that Gen3 treats datasets as completely atomic and immutable (aside from deletion), so there is no chance of one repository updating another unexpectedly via hard links.
+
+We will also designate two other non-repository subdirectories of ``/datasets`` for specific roles:
+
+ - ``/datasets/external`` holds files produced by other projects or surveys that may be of use to multiple users but does not fit into the LSST data model.  This includes the original versions of reference catalogs (e.g. Gaia DR2), truth catalogs (e.g. from DESC DC2), dust maps, etc.  Each subdirectory should have a descriptive README, and no files should be put in ``datasets/external`` itself.
+
+ - ``/datasets/testing`` holds git LFS repositories that are used in CI and rarely change.  These are provided for convenience, and should be updated when their git master branches are; there will be no attempt to make old versions available.
+
+Finally, we propose that all raw ingestion into the shared repository be done with symbolic links or in-place (outside-root) ingests, pointing to read-only filesystems - the existing ``/lsstdata`` for Rubin Observatory data, and a new ``/external-raw`` filesystem for raw data from other instruments.
+
+.. note ::
+
+   **TODO**: This document should probably say more about staging locations for (at least) raws that haven't been ingested yet, for both Rubin Observatory instruments and precursor instruments (which I imagine might be quite different in this respect).  I'd like to at least be confident that the filesystem locations and permissions I'm proposing are consistent with how data arrives and gets ingested.  I don't know enough to write that up myself.
+
```

Comment:
This leaves me wondering where raws should land on the filesystem to ensure they aren't accidentally blown away after ingestion. I also really want to know what butler hoops I need to jump through to get a cold hard FITS file (at the end of some processing pipeline, from a particular collection) in my science-hand.

Reply:
Butler doesn't need to have the raw files in the datastore at all since it can refer to the primary location in `/datasets`. This would make it impossible to delete them even if you tried. If we are using softlinks instead we would want to set permissions on that part of datastore to make it impossible for users to delete anything.

Reply:
> I also really want to know what butler hoops I need to jump through to get a cold hard FITS file (at the end of some processing pipeline, from a particular collection) in my science-hand.

Butler will at least be able to guarantee that it can get you a URI to that FITS file, assuming it exists as a [single] FITS file, and as a DM dev you'll probably know what to do with that URI.

But if you drop some of those assumptions, this becomes more of a DAX / Science Platform question, and while I'd expect the butler to be involved in the implementation of giving science users FITS files for things that maybe weren't really stored that way, it feels a little scope-creepy (to me, at least) to make it the front-end for that.

Reply:
My main concern here relates to @parejkoj's example of a science user wanting to process e.g. yesterday's DECam observations. As I understand it, if I have a new pile of last week's DECam images, I have two options: (1) ask around informally until I find somebody with permission and know-how to add things to `/datasets` and the "one giant repository" for me or (2) start my own Butler repository completely independent of everything else, which feels antithetical to the spirit of this entire document, especially if my amazing images turn out to be useful to the Project.

What I am looking for is a section of this document which improves markedly on the current state of (1), and perhaps gives guidelines for when (2) is and is not acceptable.

Reply:
If ingestion of DECam raw files is interesting to the project in general, ask @mbutlerncsa and @rgruendl about setting up automatic (transfer and) ingestion of those raw files.